File: leetcode/Python/code/1.Two-Sum/1.Two-Sum.py
"""
No.1
题目：两数之和
难度：简单

给定一个整数数组``nums``和一个目标值``target``，请你在该数组中找出和为目标值的那``两个``整数，
并返回他们的数组下标。你可以假设每种输入只会对应一个答案。但是，你不能重复利用这个数组中
同样的元素。

示例:
给定 nums = [2, 7, 11, 15], target = 9
因为 nums[0] + nums[1] = 2 + 7 = 9
所以返回 [0, 1]
"""
from typing import List


class Solution(object):
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        h = {}
        for i, j in enumerate(nums):
            k = target - j
            if k in h:
                return [h[k], i]
            h[j] = i


class Solution2(object):
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i, v in enumerate(nums):
            for j in range(i + 1, len(nums)):
                if v + nums[j] == target:
                    return [i, j]


nums = [5, 5, 5]
target = 10
d = Solution2()
e = d.twoSum(nums, target)
print(*list(e))

File: leetcode/Go/code/1.Two-Sum/1.Two-Sum.go
/*
No.1
题目：两数之和
难度：简单

给定一个整数数组``nums``和一个目标值``target``，请你在该数组中找出和为目标值的那``两个``整数，
并返回他们的数组下标。你可以假设每种输入只会对应一个答案。但是，你不能重复利用这个数组中
同样的元素。

示例:
给定 nums = [2, 7, 11, 15], target = 9
因为 nums[0] + nums[1] = 2 + 7 = 9
所以返回 [0, 1]
*/
package main

import "fmt"

func twoSumM1(nums []int, target int) []int {
	lenNums := len(nums)
	for i, v := range nums {
		for j := i + 1; j < lenNums; j++ {
			if nums[j]+v == target {
				return []int{i, j}
			}
		}
	}
	return []int{}
}

func twoSumM2(nums []int, target int) []int {
	lenNums := len(nums)
	storeMap := make(map[int]int, lenNums)

	for i, v := range nums {
		anotherNum := target - v
		if j, ok := storeMap[anotherNum]; ok {
			return []int{j, i}
		} else {
			storeMap[v] = i
		}
	}

	return []int{}
}

func main() {
	nums := []int{4, 4, 4, 4}
	target := 8
	result1 := twoSumM1(nums, target)
	result2 := twoSumM2(nums, target)
	fmt.Println("Result1:", result1)
	fmt.Println("Result2:", result2)
}
